### ISSUE TITLE
error: 'make_value_iterator' is not a member of 'pybind11'

### ISSUE THREAD

**Issue by espenhgn** [arbor-sim/arbor#1951]
**Describe the bug**

Hi! 
While building Arbor on the JSC JUSUF HPC resource, I encounter an issue with an undefined function in `pybind11`: 
```bash
...
[100%] Building CXX object python/CMakeFiles/pyarb_obj.dir/cells.cpp.o
[100%] Linking CXX executable ../bin/lmorpho
[100%] Built target lmorpho
/p/home/jusers/hagen2/jusuf/sources/arbor/python/cells.cpp: In lambda function:
/p/home/jusers/hagen2/jusuf/sources/arbor/python/cells.cpp:290:35: error: 'make_value_iterator' is not a member of 'pybind11'; did you mean 'make_key_iterator'?
  290 |                  return pybind11::make_value_iterator(ld.cache.begin(), ld.cache.end());
      |                                   ^~~~~~~~~~~~~~~~~~~
      |                                   make_key_iterator
make[2]: *** [python/CMakeFiles/pyarb_obj.dir/build.make:104: python/CMakeFiles/pyarb_obj.dir/cells.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:2017: python/CMakeFiles/pyarb_obj.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

To me, the issue appears to be that `pybind11-v2.8.1` in the git submodule should be used (CMake finds it), but the system provided `pybind11-v2.7.1` is still used during compilation. Seems like the required version was bumped via https://github.com/arbor-sim/arbor/pull/1742. 

I could get it to compile by replacing `make_value_iterator` by `make_key_iterator` in `cells.cpp`, but I don't know what sort of problems that may have introduced. 

Installation script:
```bash
module load Python/3.6.9 SciPy-Stack GCC/11.2.0 ParaStationMPI/5.5.0-1 mpi4py/3.1.3 h5py/3.5.0 GSL/2.7 git/2.33.1-nodocs CMake/3.21.1 Doxygen/1.9.1 Autotools/20210726 Boost/1.78.0 PyYAML/5.4.1

cd arbor
git checkout master
git pull
git submodule update --init --recursive

mkdir build
cd build
rm -rf *

export CC=`which gcc`
export CXX=`which g++`
cmake -DARB_WITH_PYTHON=ON   -DARB_USE_BUNDLED_LIBS=ON   -DARB_WITH_MPI=OFF   -DARB_PYTHON_LIB_PATH=~/.local/lib/python3.9/site-packages/  -DCMAKE_INSTALL_PREFIX=~/.local ..
make -j
```

CMake output:
```bash
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /p/software/jusuf/stages/2022/software/GCCcore/11.2.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /p/software/jusuf/stages/2022/software/GCCcore/11.2.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using Random123 submodule: /p/home/jusers/hagen2/jusuf/sources/arbor/ext>/random123
-- Found Python3: /p/software/jusuf/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/python3.9 (found suitable version "3.9.6", minimum required is "3.7.0") found components: Interpreter Development Development.Module Development.Embed 
-- PYTHON_EXECUTABLE: /p/software/jusuf/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/python3.9
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
Catalogue name:       dummy
Catalogue mechanisms: dummy
Extra cxx files:      
Catalogue sources:    /p/home/jusers/hagen2/jusuf/sources/arbor/test/unit/dummy
Catalogue output:     /p/home/jusers/hagen2/jusuf/sources/arbor/build/test/unit/generated/dummy
Build as standalone:  ON
Arbor cxx flags:      -march=native;$<$<BUILD_INTERFACE:$<COMPILE_LANGUAGE:CXX>>:-fvisibility=hidden>;$<$<BUILD_INTERFACE:$<COMPILE_LANGUAGE:CUDA>>:-Xcompiler=-fvisibility=hidden>
Arbor cxx compiler:   
Script prefix:        /p/home/jusers/hagen2/jusuf/sources/arbor/mechanisms
Current cxx compiler: /p/software/jusuf/stages/2022/software/GCCcore/11.2.0/bin/g++
-- Found Sphinx: /p/software/jusuf/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/sphinx-build  
-- Found PY_svgwrite: /p/home/jusers/hagen2/jusuf/.local/lib/python3.9/site-packages/svgwrite  
-- pybind11 v2.8.1 
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Python module installation path: /p/home/jusers/hagen2/jusuf/.local/lib/python3.9/site-packages
-- Configuring done
-- Generating done
-- Build files have been written to: /p/home/jusers/hagen2/jusuf/sources/arbor/build
```

**Context**

Cluster info: https://apps.fz-juelich.de/jsc/hps/jusuf/index.html
Arbor version: db887a9e4d5a986fa8f03c80dcf71eaf140588ea

**Comment by thorstenhater** [comment]
Hi @espenhgn,

thanks for reporting, that seems like a regression in our CMake indeed!

**Comment by thorstenhater** [comment]
Hi @espenhgn,

I am unable to reproduce this locally, I have pb11-2.10 installed system-wide and pb11-2.8 in a submodule.
Running as you describe reports 2.8 during configuration, just as in your output. Checking the pybind11
version (via `PYBIND11_VERSION_HEX`, but see linked PR) finds 2.8 as well. 

Could you verify that the version found is indeed the older one by using the `arbor.config()` method in 
#1961? Here's my output
```
python3 -c 'import arbor as A; print(A.__config__)'
{'mpi': False, 'mpi4py': False, 'gpu': False, 'vectorize': False, 'profiling': False, 'neuroml': False, 'bundled': True, 'version': '0.7.1-dev', 'source': '2022-08-25T17:32:16+02:00 caa9a0b0ce08d71b8e6a99b24d0f66b780ae2c97 modified', 'arch': 'native', 'pybind-version': '2.8.1'}
```

**Comment by brenthuisman** [comment]
Heads up: if you've checkout out Arbor fresh, last week Pybind was bumped again: https://github.com/arbor-sim/arbor/commit/e67347b61e054aa7cbafd092211de85994375421

I don't know if it's a typo, but you're loading the `Python/3.6.9` module, but then I find traces of Python 3.9(.6) in your CMake output.

**Comment by espenhgn** [comment]
Hi @brenthuisman ,
Thanks for testing! I assume you did this locally? 

I now think the explanation is a `sys.path` ordering issue resulting from the `module load Python/3.9.6 ...` above. 
This prepends a bunch of things to Python's search path, while `~ /.local/lib/python3.9/site-packages` is by default further down the list. Consequently, the system-provided pybind11 is found first. 

The possible workarounds for now appear to be (1) don't rely on any of the provided packages (at least nothing that would add pybind11 as a dependency), (2) create a virtualenv for Arbor stuff, or (3) install&run Arbor in a container. 

I think this issue can be listed as a low priority then. I'll probably enquire with the JSC support team, as it is rather an issue with their setup.

**Comment by thorstenhater** [comment]
Note to us: Maybe our use of CMake/Pybind11 is wrong. See this for a simple way that might even be more correct:
https://pybind11.readthedocs.io/en/stable/compiling.html#building-with-cmake

**Comment by brenthuisman** [comment]
Use CMake's `pybind11_add_module` you mean?

**Comment by brenthuisman** [comment]
> Hi @brenthuisman , Thanks for testing! I assume you did this locally?

Nope, I just read your log :)

> I now think the explanation is a `sys.path` ordering issue resulting from the `module load Python/3.9.6 ...` above. This prepends a bunch of things to Python's search path, while `~ /.local/lib/python3.9/site-packages` is by default further down the list. Consequently, the system-provided pybind11 is found first.

Invoking `-DARB_USE_BUNDLED_LIBS=ON` should cause CMAKE not to search for Pybind11, but grab it from the project source. I don't see how appending/modifying paths would influence that. What am I missing?

#1968 _actually_ uses `pb11_src_dir` for `add_subdirectory()`. Not sure if that really matters, but since this is CMAKE... Can you try?

**Comment by espenhgn** [comment]
> Invoking `-DARB_USE_BUNDLED_LIBS=ON` should cause CMAKE not to search for Pybind11, but grab it from the project source. I don't see how appending/modifying paths would influence that. What am I missing?

I don't think you missed anything. 

Anyway; did some testing. With no system-provided pybind11 in the path, from requesting a subset of modules compilation works (tested on JURECA; JUSUF is in maintenance) using the current master (SHA: 8f5259143e6c1d6fd9a0991ca1bfff0742f04ca8).
```
module purge
module load Stages/2022 GCCcore/.11.2.0 Python/3.9.6 SciPy-Stack/2021b CMake/3.21.1
export CC=`which gcc`
export CXX=`which g++`
cmake -DARB_WITH_PYTHON=ON   -DARB_USE_BUNDLED_LIBS=ON   -DARB_WITH_MPI=OFF   -DARB_PYTHON_LIB_PATH=~/.local/lib/python3.9/site-packages/  -DCMAKE_INSTALL_PREFIX=~/.local ..
make -j
make install
```
succeeds. 

Config info:
```
$ python3 -c 'import arbor as A; print(A.__config__)'
{'mpi': False, 'mpi4py': False, 'gpu': False, 'vectorize': False, 'profiling': False, 'neuroml': False, 'bundled': True, 'version': '0.7.1-dev', 'source': '2022-09-06T10:20:07+02:00 8f5259143e6c1d6fd9a0991ca1bfff0742f04ca8', 'arch': 'native'}
```

As for the changes in https://github.com/arbor-sim/arbor/pull/1968, these are most likely fine, but they don't fix the issue I encountered unfortunately, when an older system-provided pybind11 version is available.

**Comment by brenthuisman** [comment]
Does it help if you also pass the correct (to be abundantly certain) `PYTHON_EXECUTABLE` to CMake?

**Comment by espenhgn** [comment]
Hi again. I tried passing `PYTHON_EXECUTABLE` to `CMake` and `VERBOSE=1` to `make`. 
`CMake` output seems ok:
```
$ cmake -DARB_WITH_PYTHON=ON   -DARB_USE_BUNDLED_LIBS=ON   -DARB_WITH_MPI=OFF   -DARB_PYTHON_LIB_PATH=~/.local/lib/python3.9/site-packages/  -DCMAKE_INSTALL_PREFIX=~/.local -DPYTHON_EXECUTABLE=/p/software/jurecadc/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/python ..
Re-run cmake no build system arguments
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /p/software/jurecadc/stages/2022/software/GCCcore/11.2.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /p/software/jurecadc/stages/2022/software/GCCcore/11.2.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using Random123 submodule: /p/project/cjinb33/hagen1/sources/arbor/ext>/random123
-- Found Python3: /p/software/jurecadc/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/python (found suitable version "3.9.6", minimum required is "3.7.0") found components: Interpreter Development Development.Module Development.Embed 
-- PYTHON_EXECUTABLE: /p/software/jurecadc/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/python
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
Catalogue name:       dummy
Catalogue mechanisms: dummy
Extra cxx files:      
Catalogue sources:    /p/project/cjinb33/hagen1/sources/arbor/test/unit/dummy
Catalogue output:     /p/project/cjinb33/hagen1/sources/arbor/build/test/unit/generated/dummy
Build as standalone:  ON
Arbor cxx flags:      -march=native;$<$<BUILD_INTERFACE:$<COMPILE_LANGUAGE:CXX>>:-fvisibility=hidden>;$<$<BUILD_INTERFACE:$<COMPILE_LANGUAGE:CUDA>>:-Xcompiler=-fvisibility=hidden>
Arbor cxx compiler:   
Script prefix:        /p/project/cjinb33/hagen1/sources/arbor/mechanisms
Current cxx compiler: /p/software/jurecadc/stages/2022/software/GCCcore/11.2.0/bin/g++
-- Found Sphinx: /p/software/jurecadc/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/bin/sphinx-build  
-- Could NOT find PY_svgwrite (missing: PY_SVGWRITE) 
-- pybind11 v2.10.0 
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Python module installation path: /p/home/jusers/hagen1/jureca/.local/lib/python3.9/site-packages
-- Configuring done
-- Generating done
-- Build files have been written to: /p/project/cjinb33/hagen1/sources/arbor/build
```
`make VERBOSE=1` fails at
```
...
[ 91%] Building CXX object python/CMakeFiles/pyarb_obj.dir/cells.cpp.o
cd /p/project/cjinb33/hagen1/sources/arbor/build/python && /p/software/jurecadc/stages/2022/software/GCCcore/11.2.0/bin/g++  -I/p/project/cjinb33/hagen1/sources/arbor/arbor/include -I/p/project/cjinb33/hagen1/sources/arbor/build/arbor/include -I/p/project/cjinb33/hagen1/sources/arbor/arborenv/include -I/p/project/cjinb33/hagen1/sources/arbor/build/arborenv/include -I/p/project/cjinb33/hagen1/sources/arbor/arborio/include -I/p/project/cjinb33/hagen1/sources/arbor/build/arborio/include -isystem /p/project/cjinb33/hagen1/sources/arbor/ext/pybind11/include -isystem /p/software/jurecadc/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/include/python3.9 -O3 -DNDEBUG -fPIC -fvisibility=hidden -Wall -Wno-maybe-uninitialized -Wno-psabi -std=c++17 -MD -MT python/CMakeFiles/pyarb_obj.dir/cells.cpp.o -MF CMakeFiles/pyarb_obj.dir/cells.cpp.o.d -o CMakeFiles/pyarb_obj.dir/cells.cpp.o -c /p/project/cjinb33/hagen1/sources/arbor/python/cells.cpp
/p/project/cjinb33/hagen1/sources/arbor/python/cells.cpp: In lambda function:
/p/project/cjinb33/hagen1/sources/arbor/python/cells.cpp:299:35: error: 'make_value_iterator' is not a member of 'pybind11'; did you mean 'make_key_iterator'?
  299 |                  return pybind11::make_value_iterator(ld.cache.begin(), ld.cache.end());
      |                                   ^~~~~~~~~~~~~~~~~~~
      |                                   make_key_iterator
make[2]: *** [python/CMakeFiles/pyarb_obj.dir/build.make:104: python/CMakeFiles/pyarb_obj.dir/cells.cpp.o] Error 1
make[2]: Leaving directory '/p/project/cjinb33/hagen1/sources/arbor/build'
make[1]: *** [CMakeFiles/Makefile2:2065: python/CMakeFiles/pyarb_obj.dir/all] Error 2
make[1]: Leaving directory '/p/project/cjinb33/hagen1/sources/arbor/build'
make: *** [Makefile:136: all] Error 2
```
where `-isystem /p/project/cjinb33/hagen1/sources/arbor/ext/pybind11/include` is listed prior to `-isystem /p/software/jurecadc/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/include/python3.9`. 

There's no `-I/p/project/cjinb33/hagen1/sources/arbor/ext/pybind11/include` that I can see (but maybe I should not expect that). 

Should we persevere with this issue as the workaround is quite straightforward (unloading modules that may load pybind11 while building Arbor)?

**Comment by thorstenhater** [comment]
Here's the culprit: SciPy-bundle, loaded via SciPy-Stack and is setting this
```
whatis("Description: Bundle of Python packages for scientific software")
whatis("Homepage: https://python.org/")
whatis("URL: https://python.org/")
whatis("Extensions: beniget-0.4.1, Bottleneck-1.3.2, deap-1.3.1, gast-0.5.2, mpmath-1.2.1, numexpr-2.7.3, numpy-1.21.3, pandas-1.3.4, ply-3.11, pythran-0.10.0, scipy-1.7.1")
conflict("SciPy-bundle")
load("imkl/.2021.4.0")
load("Python/3.9.6")
load("pybind11/.2.7.1")
prepend_path("CMAKE_PREFIX_PATH","/p/software/jusuf/stages/2022/software/SciPy-bundle/2021.10-gcccoremkl-11.2.0-2021.4.0")
prepend_path("LIBRARY_PATH","/p/software/jusuf/stages/2022/software/SciPy-bundle/2021.10-gcccoremkl-11.2.0-2021.4.0/lib")
prepend_path("PATH","/p/software/jusuf/stages/2022/software/SciPy-bundle/2021.10-gcccoremkl-11.2.0-2021.4.0/bin")
setenv("EBROOTSCIPYMINBUNDLE","/p/software/jusuf/stages/2022/software/SciPy-bundle/2021.10-gcccoremkl-11.2.0-2021.4.0")
setenv("EBVERSIONSCIPYMINBUNDLE","2021.10")
setenv("EBDEVELSCIPYMINBUNDLE","/p/software/jusuf/stages/2022/software/SciPy-bundle/2021.10-gcccoremkl-11.2.0-2021.4.0/easybuild/Compiler-GCCcore-11.2.0-SciPy-bundle-2021.10-easybuild-devel")
prepend_path("PYTHONPATH","/p/software/jusuf/stages/2022/software/SciPy-bundle/2021.10-gcccoremkl-11.2.0-2021.4.0/lib/python3.9/site-packages")
setenv("EBEXTSLISTSCIPYMINBUNDLE","numpy-1.21.3,ply-3.11,gast-0.5.2,beniget-0.4.1,pythran-0.10.0,scipy-1.7.1,numexpr-2.7.3,Bottleneck-1.3.2,pandas-1.3.4,mpmath-1.2.1,deap-1.3.1")
```
critically `CMAKE_PREFIX_PATH` is being set, which might be the issue and also why I cannot reproduce this locally, since
I do not use modules.

For now, yes, just unload SciPy-Stack until the build is through and will figure out what's wrong.

**Comment by thorstenhater** [comment]
Soooo, I tracked that bastard of a bug down and it's neither ours nor can we fix it. Unloading Scipy-Stack is the only way.
What happens is that the `module` commnad writes into `CPATH` which is a variable read by GCC that contains directories
used _by all compilations, implicitly_. CMake can do all it want, GCC will use these, as specified by GCC's docs.
Thus running `gcc -M` to make dependencies for a C++ file spits out the wrong header.

Module (and easybuild, too) use `CPATH` to free you from the hassle of configuring your builds with millions of `-I` flags.
It's a very big hammer and it hit us hard. So, yeah, until the systems have pybind11-2.9+, just unload that module.

**Comment by thorstenhater** [comment]
I'll drop a line in the build docs.

**Comment by brenthuisman** [comment]
[This](https://github.com/easybuilders/easybuild-framework/issues/3331) implies there is an order, and `-I` takes precedence over `CPATH`. Since (AFAIK) CMake is setting `-I` flangs, this should not be it?

Separately, seems like setting CPATH is bad behavior according to some/many. Where to we file change requests for JSC infra?

**Comment by thorstenhater** [comment]
Quoting myself from above:
> Thus running `gcc -M` to make dependencies for a C++ file spits out the wrong header.

**Comment by thorstenhater** [comment]
Well, it's not directly JSC's doing. They use EasyBuild and there we find
https://github.com/easybuilders/easybuild-framework/issues/3331

**Comment by brenthuisman** [comment]
That does not seem an answer; if `-I` has precedence, why would pybind11 as set by CMake not be found first? Usually these things stop resolving after a first encounter.

**Comment by thorstenhater** [comment]
It's found by CMake, as seen in the logs, that's part of the problem: We find and report the correct version, but **include**
the wrong one. CMake uses `g++ -M` and **`GCC`** finds `$CPATH` and picks that first.

**Comment by brenthuisman** [comment]
Ah. `-M`. `module` and `gcc` are giving me lots of hits on those other (C++) modules. Does anyone know if this CMake specific? Would Meson give the same issue?

**Comment by thorstenhater** [comment]
So, here's what I did:
- copy out the `g++` line that's failing
- you'll get this by `make VERBOSE=1`
- remove all the options like `-MT` and similar plus their args
- add `-M`
- pipe the command into a file
- browse to find where pybind11 is coming from

**Comment by thorstenhater** [comment]
Hey,

we have just merged #1961 which denies building with incompatible Pybind11 and this issue has work-around
identified. I am closing this, if still problematic, please reopen.
